### PR TITLE
feat: builds namespace

### PR DESCRIPTION
**Pull request by vladfrangu**
Screenshots incoming, tests too (probably will rely on cucumber tests)

Closes part of the #619 epic

<details>
<summary>builds ls</summary>

Code - 2024-08-26 at 19 49 20

</details>

<details>
<summary>builds info</summary>

Code - 2024-08-26 at 19 49 53

</details>

<details>
<summary>builds info --json</summary>

Code - 2024-08-26 at 19 51 30

</details>

<details>
<summary>builds create --actor=id</summary>

Code - 2024-08-26 at 20 33 22

</details>

<details>
<summary>builds create --actor=id --tag=string | multiple versions tagged the same way</summary>

Code - 2024-08-26 at 20 34 22

</details>

<details>
<summary>builds create --actor=id --tag=non-existent</summary>

THIS NEEDS A REVIEW FROM @netmilk!!

Code - 2024-08-26 at 20 35 11

</details>

<details>
<summary>builds create --actor=id --tag=string --version=string</summary>

Code - 2024-08-26 at 20 35 37

</details>

### CODE REVIEW

**Comment by netmilk** [comment]
@vladfrangu I've manually validated all the functionality provided by this PR, and it works as expected. Thank you! I'd wait with merging for the Cucumber spec, though.

Just one improvement proposal:
Shouldn't the `$ apify build create` always attach to the log the same way `apify push` does? If so, I'd add `--detach` flag to exit `0` right after successfully creating a build.

**Comment by netmilk** [comment]
I've just noticed the `apify builld log` subcommand, which is implemented and isn't mentioned in this PR. Because of that I've missed it initially.

**Comment by vladfrangu** [comment]
Thats what it does @netmilk, re `apify build log` 👀 

Related to `apify builds create`, do we.. want to also attach the logs automatically? Can we get that documented on notion too (also maybe make it opt-in, or bring in the --wait-for-finish flag from other cmds)

**Comment by netmilk** [comment]
Understood. Perfect. Thank you for explaining it to me.

What about `$ apify builds create --log` and let's keep attach / dettach words for the streaming?

**Comment by vladfrangu** [comment]
So what would --log do? Attach the log stream on command run? And wdym by attach / detach for streaming?

**Comment by netmilk** [comment]
@vladfrangu, I'm sorry for the ambiguity. Yes, it would connect `apify build log` to the newly created build. 

By "streaming" I mean connecting to platform live events during a `run`. Run, Dataset, KVS, RQ Creation, Status messages, push-data, set-value [as briefly mentioned in the internal spec](https://www.notion.so/apify/API-CLI-design-Final-Draft-517b46afd5b8431dae7e38b06c9080fc?pvs=4#accad62cee8e4121b4d960f41427c427).

**Review by B4nan — CHANGES_REQUESTED** [review]
looks good, but we need at least the cucumber tests before merging

File: src/commands/builds/ls.ts
Comment (vladfrangu) [reply src/commands/builds/ls.ts:123]: If anything its a good callout that we should extract that to a variable and not inline it

File: src/lib/commands/pretty-print-status.ts
Comment (vladfrangu) [reply src/lib/commands/pretty-print-status.ts:1]: Yop. Idfk how it got up there, but it wasn't supposed to be up there

File: src/lib/commands/pretty-print-bytes.ts
Comment (vladfrangu) [reply src/lib/commands/pretty-print-bytes.ts:1]: It's meant to represent utilities for commands, or anything _internal_ to what commands do/need to do (see runActorOnCloud)

File: src/commands/builds/create.ts
Comment (barjin) [inline src/commands/builds/create.ts:15]:
```suggestion
				'Optional Actor ID or Name to trigger a build for. By default, it will use the Actor from the current directory.',
```

File: src/commands/builds/create.ts
Comment (barjin) [inline src/commands/builds/create.ts:18]:
```suggestion
			description: 'Build tag to be applied to the successful Actor build. By default, this is "latest".',
```

File: src/commands/builds/create.ts
Comment (barjin) [inline src/commands/builds/create.ts:99]:
```suggestion
				message: `No Actor versions with the tag "${tag}" found. You can push a new version with this tag by using "apify push --build-tag=${tag}".`,
```

File: src/commands/builds/info.ts
Comment (barjin) [inline src/commands/builds/info.ts:11]:
```suggestion
	static override description = 'Prints information about a specific build.';
```

File: src/commands/builds/info.ts
Comment (barjin) [inline src/commands/builds/info.ts:16]:
```suggestion
			description: 'The build ID to get information about.',
```

File: src/commands/builds/info.ts
Comment (barjin) [inline src/commands/builds/info.ts:30]:
```suggestion
			error({ message: `Build with ID "${buildId}" was not found on your account.` });
```

File: src/commands/builds/log.ts
Comment (barjin) [inline src/commands/builds/log.ts:13]:
```suggestion
			description: 'The build ID to get the log from.',
```

File: src/commands/builds/log.ts
Comment (barjin) [inline src/commands/builds/log.ts:8]:
```suggestion
	static override description = 'Prints the log of a specific build.';
```

File: src/commands/builds/log.ts
Comment (barjin) [inline src/commands/builds/log.ts:25]:
```suggestion
			error({ message: `Build with ID "${buildId}" was not found on your account.` });
```

File: src/commands/builds/ls.ts
Comment (barjin) [inline src/commands/builds/ls.ts:41]:
```suggestion
				'Optional Actor ID or Name to list builds for. By default, it will use the Actor from the current directory.',
```

File: src/commands/builds/ls.ts
Comment (barjin) [inline src/commands/builds/ls.ts:74]:
```suggestion
				message: `${ctx.reason}. Please run this command in an Actor directory, or specify the Actor ID by running this command with "--actor=<id>".`,
```

**Review by barjin — COMMENTED** [review]
Just a few copy nits I noticed when skimming the PR... the screenshots look bomb!

Btw do we communicate that "Builds" are a platform-only feature anywhere? As a new user, I would definitely assume that I can build the Actors locally with CLI as well (maybe more of a question for @netmilk , sorry if this has been discussed somewhere before).

File: src/commands/builds/ls.ts
Comment (janbuchar) [inline src/commands/builds/ls.ts:64]: Please split the method into smaller chunks.

File: src/commands/builds/info.ts
Comment (janbuchar) [inline src/commands/builds/info.ts:81]: Why though? Is the type missing in client/spec? Or is it just free-form data?

File: src/commands/builds/info.ts
Comment (janbuchar) [inline src/commands/builds/info.ts:74]: Is there any way we could make this more readable? The amount of intertwined ternaries and template strings here is staggering.

**Review by janbuchar — COMMENTED** [review]
I noticed that your `run` methods tend to be very long, which makes them hard to read. Other than that and a few minor comment, I think this is good to go.

File: .github/workflows/cucumber.yaml
Comment (vladfrangu) [reply .github/workflows/cucumber.yaml:8]: Yop, external contributors editing cucumber files could trigger a run with our test user token. Technically it should be safe still but I don't wanna risk it. IIRC github already requires approval for new contributors, but if you guys have a way to always require approval for external contributors please link me 🙏

File: src/commands/builds/ls.ts
Comment (vladfrangu) [reply src/commands/builds/ls.ts:64]: This request is too vague. Why? (And don't look at call or run 😓, those predate me but I have been wanting to rewrite and shorten them for a while now)

What even is there to split into smaller chunks? Would love some callouts so I can understand better

File: src/commands/builds/info.ts
Comment (vladfrangu) [reply src/commands/builds/info.ts:81]: apify-client types are a mess (missing types, wrong types, i-was-stupid-when-writing-types types), and iirc haven't been updated in *ages*. Hopefully one day we can do it more automated 😬

File: src/commands/builds/ls.ts
Comment (janbuchar) [reply src/commands/builds/ls.ts:64]: That is a tough question to answer for someone who didn't write this :smile: 

My reasoning is that there is a lot of different variables in use (~20?), and the human brain can only hold like five "concepts" at once. If you have small functions where it's clear what goes in and what goes out, you can restructure the `run` method so that you can tell at a glance what the steps are and what's there output, and you can also zoom in into each step and check all the nasty details.

File: .github/workflows/cucumber.yaml
Comment (janbuchar) [reply .github/workflows/cucumber.yaml:8]: https://github.com/apify/apify-sdk-python/blob/master/.github/workflows/run_checks.yaml#L21-L49

File: src/commands/builds/info.ts
Comment (janbuchar) [reply src/commands/builds/info.ts:81]: Gotcha. Then I'd prefer to open an issue in the client repo and link it in that comment. Also, we have an openapi spec, and if we ever want automated client generation, it'd be great if you could check if the type is correct there at least.

File: src/commands/builds/info.ts
Comment (janbuchar) [reply src/commands/builds/info.ts:74]: That would help!

File: src/commands/builds/info.ts
Comment (vladfrangu) [reply src/commands/builds/info.ts:81]: https://github.com/apify/apify-client-js/issues/526 exists for tracking that already 😅

Comment (janbuchar) [reply src/commands/builds/info.ts:81]: Perfect! Is there a reason not to link that issue from the code? Or at least make the issue a TODO? Either of those options would make it easier to search for...

Comment (vladfrangu) [reply src/commands/builds/info.ts:81]: No particular reason, I'll update them to TODOs and link the issue

File: src/commands/builds/ls.ts
Comment (vladfrangu) [reply src/commands/builds/ls.ts:65]: I tried some splitting up, but unfortunately I don't know how else to split it up more than just how it is right now 😅

File: .github/workflows/cucumber.yaml
Comment (vladfrangu) [reply .github/workflows/cucumber.yaml:8]: Ty! I'll tackle this later 🙏

File: src/commands/builds/ls.ts
Comment (janbuchar) [reply src/commands/builds/ls.ts:65]: Well, such is life. No need to stall the PR then, but could we make an issue to track this and possible all the other instances of loooong `run` methods?

Comment (vladfrangu) [reply src/commands/builds/ls.ts:65]: Sure, but its probably gonna end up in the icebox of the icebox 😅

**Review by B4nan — APPROVED** [review]

**Comment by B4nan** [comment]
as discussed on slack, i would love to see more e2e test assertions, i get that it might be too complicated to deal with it properly, but at least some lower effort asserts (e.g. to detect one line we know should be there) would be great

**Pull request merged**
